arch/contracts/exchange/program/src/lib.rs

```
@@ -0,0 +1,429 @@
+pub struct InitProgramStateParams {
+    pub fee_account: String,
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct InitTokenStateParams {
+    pub token_id: String,
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct InitWalletBalancesParams {
+    pub token_balance_setups: Vec<TokenBalanceSetup>,
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct DepositBatchParams {
+    pub token_deposits: Vec<TokenDeposits>,
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct TokenDeposits {
+    pub account_index: u8,
+    pub deposits: Vec<Adjustment>,
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct TokenWithdrawals {
+    pub account_index: u8,
+    pub fee_address_index: u32,
+    pub withdrawals: Vec<Withdrawal>,
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct WithdrawBatchParams {
+    pub token_withdrawals: Vec<TokenWithdrawals>,
+    pub tx_hex: Vec<u8>,
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct SettlementAdjustments {
+    pub account_index: u8,
+    pub increments: Vec<Adjustment>,
+    pub decrements: Vec<Adjustment>,
+    pub fee_amount: u64,
+}
+
+#[derive(Clone, BorshSerialize, BorshDeserialize)]
+pub struct SettlementBatchParams {
+    pub settlements: Vec<SettlementAdjustments>
+}
+
+
+pub fn init_program_state(accounts: &[AccountInfo],
+                          params: InitProgramStateParams) -> Result<(), ProgramError> {
+
+    let state_data: Vec<u8> = get_account_data(accounts, 0)?;
+    let mut state: ProgramState = if state_data.is_empty() {
+        ProgramState {
+            version: 0,
+            fee_account_address: "".to_string(),
+            settlement_batch_hash: "".to_string(),
+            last_settlement_batch_hash: "".to_string(),
+            last_withdrawal_batch_hash: "".to_string(),
+        }
+    } else {
+        borsh::from_slice(&state_data).unwrap()
```

Should we return an error here perhaps? Or was it a conscious decision to allow program state initialization for than once?

---

Its probably better as you suggest to fail here if they try to re-initialize - the original idea was to be able to update the fee account address but if we need to do that better to have an explicit means - I will add a ticket to fix this